### DISCUSSION TITLE
Variable first UX

### DISCUSSION THREAD

**Discussion by adrientetar** [fontra/fontra#534] [General]
Hi! I saw the atypi talk slides for Fontra, and I'm curious about the "variable first" editor concept.

Does it mean to have more capabilities than in the "multiple master" interpolation model?

I think the only related concept i have seen is Simon Cozens masterless editing: https://twitter.com/simoncozens/status/1390338924984348677
but that could get a little messy if creating masters at arbitrary places for different glyphs.

**Comment by justvanrossum** [comment]
All it was meant to mean is that the application is designed with interpolation/variability in mind from the start, and not as a feature bolted onto a more static editor concept.

Regarding interpolation models: for better or for worse, Fontra follows the OpenType VF model quite closely, in that it adapts fontTools.varLib's VariationModel. Although we're looking into expanding that somewhat with extrapolation (#418).

**Comment by davelab6** [comment]
@simoncozens also wrote more about this at https://simoncozens.github.io/colrv1-rocks

I believe though that this editor-concept goes back to the origin of VF, with the "RoyalT" (or "Mutator"?) editor developed within Apple 30 years ago as part of the original GX effort: It does indeed mean having more capabilities than in the "multiple master" interpolation model, because the design space becomes both easier to deal with as there are a lot more interpolated/inferred "corners", and harder, as you say, because it becomes more complex (and  'messy'). But that's where the power of this model comes from, and dealing with such complex design spaces is a UI/UX problem that I believe is soluble.

**Comment by davelab6** [comment]
@felipesanches pointed out that the old school Apple tools are on Wayback Machine! And @lorp found https://web.archive.org/web/19980207053029/http://fonts.apple.com/ and said, "Just click the tool you want, accept the license, then go to a page that lets you download them all." 

I see it works to drag and drop the attached into https://system7.app - there, RoyalT runs and can open the system's Helvetica font, but, mutator will not run... and RoyalT bombs out on most of the system fonts. @twardoch pointed out http://macintoshgarden.org/apps/ttconverter can run there to convert TTFs to the resource-fork data needed by these apps.

Proofing such large and intricate design spaces is a challenge, for which @dberlow designed https://videoproof.appspot.com

**Comment by dberlow** [comment]
Try opening Skia?On Jun 27, 2023, at 6:00 PM, Dave Crossland ***@***.***> wrote:﻿
@felipesanches pointed out that the old school Apple tools are on Wayback Machine! And @Lorp found https://web.archive.org/web/19980207053029/http://fonts.apple.com/ and said, "Just click the tool you want, accept the license, then go to a page that lets you download them all."
I see it works to drag and drop the attached into https://system7.app - there, RoyalT runs and can open the system's Helvetica font, but, mutator will not run... and RoyalT bombs out on most of the system fonts. @twardoch pointed out http://macintoshgarden.org/apps/ttconverter can run there to convert TTFs to the resource-fork data needed by these apps.
Proofing such large and intricate design spaces is a challenge, for which @dberlow designed https://videoproof.appspot.com

**Comment by Lorp** [comment]
Tried it. Crashed the OS.

**Comment by justvanrossum** [comment]
FWIW.

https://github.com/googlefonts/fontra/assets/4246121/25f25e69-f9ef-4543-b349-4db2f70f526e

**Comment by dberlow** [comment]
Did you call customer support at Apple?On Jun 28, 2023, at 7:32 AM, Laurence Penney ***@***.***> wrote:﻿
Tried it. Crashed the OS.

**Comment by adrientetar** [comment]
> FWIW.

Cool thanks for sharing! nice work on the editor.

As I understand, the "layers" are just like keyframes and you have sliders to adjust the preview globally among the previewed text. And uncheck "on" checkbox to exclude the layer from interpolation? (what's the difference between "on" and "bg"?)

If one of the glyphs in the preview only has one master, they will just not interpolate?
I assume, if you put the slider at a position without a keyframe and start moving points, it saves a new "keyframe" at that position?
Do you define the font metrics together with the axes in a "font info" window? Since these need to be interpolated as well and should always be present.

That is pretty cool. I guess though, for the majority of fonts, designers will probably stick to the multiple master model, I mean like make a Regular weight, draw all the glyphs, then derive a Bold weight and draw all the glyphs or something like that. But probably this type of UI is more fluid or flexible and you can tweak the position of the static glyph instances more easily?

**Reply by justvanrossum** [reply]
Our UI is for sure not final, and we do want to make it easier to start drawing at a designspace location where the glyph doesn't have a source yet.

In general, I'd expect all corners of a designspace to get filled. But whether a glyph needs intermediate sources, and where, could be decided on a per-glyph basis.

> Do you define the font metrics together with the axes in a "font info" window? Since these need to be interpolated as well and should always be present.

Font metrics is still uncharted territory in Fontra. But yeah, I'd like a similar mechanism as what we do with glyphs: you would only need to fill in the values for those locations where you need them to vary.

**Comment by BlackFoundry** [comment]
Uncheck "on" will indeed exclude the source from interpolation. The "bg" checkbox will display the source in the background (when there are many sources you may not want to see them all together in the background)

> I assume, if you put the slider at a position without a keyframe and start moving points, it saves a new "keyframe" at that position?

As of today a dialog will pop up saying "cannot edit, location is not a source"; but we are planing to improve this dialog soon, see https://github.com/googlefonts/fontra/issues/591

> Do you define the font metrics together with the axes in a "font info" window? Since these need to be interpolated as well and should always be present.

In the right sidebar there is a field to edit the width of the source
